### PR TITLE
Add an endpoint to return possible kubelet versions inside the cluster

### LINKED ISSUE #1386: Write endpoint to return possible kubelet versions for a cluster
### Current state
Each node which gets created, gets created with the same version as the master is at that moment.
Though the add-node endpoint already offers to specify a version.
The UI does not set the version at the moment as theres no clear definition in what versions are possible.

### Desired state
The kubermatic api exposes an endpoint which returns all possible versions for nodes.
Respect version of the cluster & the allowed node versions from https://github.com/kubermatic/kubermatic/blob/master/config/kubermatic/static/master/versions.yaml

### PR DESCRIPTION
**Pull request by maciaszczykm**
**What this PR does / why we need it**: Adds an endpoint to return possible kubelet versions inside the cluster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: Fixes #1386.

**Special notes for your reviewer**:

**Documentation**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: api/pkg/handler/upgrade.go [hunk 1]
```diff
@@ -47,21 +49,68 @@ func getClusterUpgrades(updateManager UpdateManager, projectProvider provider.Pr
...
 
-		sv := make([]*apiv1.MasterVersion, len(versions))
-		for v := range versions {
-			sv[v] = &apiv1.MasterVersion{
-				Version: versions[v].Version,
-				Default: versions[v].Default,
+		var compatibleVersions []*version.MasterVersion
+		clusterVersion := cluster.Spec.Version.Semver()
+		for _, v := range versions {
+			if err = ensureVersionCompatible(clusterVersion, v.Version); err == nil {
```
Comment (p0lyn0mial) [inline api/pkg/handler/upgrade.go:80]: I think we could simplify the code, how about ?

```
 if err = ensureVersionCompatible(clusterVersion, v.Version); err == nil {
  compatibleVersions = append(compatibleVersions, v)
  continue
 } 
 _, ok := err.(errVersionSkew)
 if !ok {
  return nil, fmt.Errorf("failed to check compatibility between kubelet %q and control plane %q: %v", v.Version, clusterVersion, err)
 } 
```

File: api/pkg/handler/routes_v1.go [hunk 2]
```diff
@@ -142,6 +142,10 @@ func (r Routing) RegisterV1(mux *mux.Router) {
 		Path("/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/upgrades").
 		Handler(r.getClusterUpgrades())
 
+	mux.Methods(http.MethodGet).
+		Path("/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/versions").
```
Comment (p0lyn0mial) [inline api/pkg/handler/routes_v1.go]: there is an endpoint for returning all possible updates for a cluster under `projects/g9grmq8kp7/dc/europe-west3-c/clusters/lfczms4mh5/upgrades`. The endpoint you added is for machines/nodes, thus can we move it to `/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/nodes/upgrades` ?

File: api/pkg/handler/upgrade_test.go [hunk 3]
```diff
@@ -121,3 +121,125 @@ func TestGetClusterUpgradesV1(t *testing.T) {
...
+func TestGetClusterVersions(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name                   string
+		cluster                *kubermaticv1.Cluster
+		existingKubermaticObjs []runtime.Object
+		apiUser                apiv1.User
+		versions               []*version.MasterVersion
+		updates                []*version.MasterUpdate
+		compatible             []*apiv1.MasterVersion
```
Comment (p0lyn0mial) [inline api/pkg/handler/upgrade_test.go]: can we rename it to `expectedOutput` ?

File: api/pkg/handler/upgrade_test.go [hunk 4]
```diff
@@ -121,3 +121,125 @@ func TestGetClusterUpgradesV1(t *testing.T) {
...
 }
+
+func TestGetClusterVersions(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name                   string
+		cluster                *kubermaticv1.Cluster
+		existingKubermaticObjs []runtime.Object
+		apiUser                apiv1.User
+		versions               []*version.MasterVersion
```
Comment (p0lyn0mial) [inline api/pkg/handler/upgrade_test.go]: can we rename it to `existingVersions` ?

File: api/pkg/handler/upgrade_test.go [hunk 5]
```diff
@@ -121,3 +121,125 @@ func TestGetClusterUpgradesV1(t *testing.T) {
...
+		updates                []*version.MasterUpdate
```
Comment (p0lyn0mial) [inline api/pkg/handler/upgrade_test.go]: can we please rename it to `existingUpdates` ?

**Comment by p0lyn0mial** [comment]
/lgtm

waiting for @kdomanski before approval

**Comment by p0lyn0mial** [comment]
@maciaszczykm since our API is not public we can remove the release note. See https://github.com/kubermatic/kubermatic/pull/2676#issuecomment-461403440

**Comment by maciaszczykm** [comment]
@p0lyn0mial Done 👍

**Pull request merged**